docs/RFCS/20211013_contentions.md

```diff
@@ -0,0 +1,393 @@
+timeseries data with a timsstamp as its first column.  Each row in the table
+represents an instance of contention event from the perspective of the blocked
+transaction.
+
+The gateway node is responsible for recording contention events for the blocked
+transaction it executed. Most of the data listed in the table are readily
+available to the gateway node today.
+
+`blocked_txn_id` can be directly fetched from within the `connExecutor` at the
+gateway node and the `contending_txn_id` can be fetched from the tracing spans.
+
+`blocked_stmt_fingerprint_id` and `blocked_txn_fingerprint_id` allow the
+new `contended_queries` table to be joined into the existing SQL Stats tables
+and is already available from the SQL Stats subsystem. `contention_duration`
+and `contented_row` is already available to us in the contention event proto.
+
+The challenging part of this new system is to populate
+`contending_txn_fingerprint_id` column. Effectively, this means we need a
+mechanism to resolve a transaction ID (that belongs to a transaction that was
+potentially executed on a different node) into its corresponding a transaction
+fingerprint ID.
+
+## Transaction ID Resolution
+
+If the contending transaction was executed on the same node where blocked
+transaction was executed, then Transaction ID Resolution is trivially solvable.
+However, this is often not the case.
+
+Since it is entirely possible that the contending transaction was executed on a
+different node, the gateway node that executed the blocked transaction needs to
+contact a remote node to perform Transaction ID resolution. There are few
+options on how this remote Transaction ID resolution can be implemented.
+Each one of the option requires a different degree of involvement from the KV
+team.
+
+### Option 1: No KV Team Involvement
+
+In this scenario, SQL Observability will proceed to implement Transaction ID
+resolution without any involvement from KV Team. The detailed technical design
+is as follow:
+
+Each node in the CockroachDB maintains an in-memory FIFO cache of recently
+executed transactions. (namely `TxnIDCache`). The in-memory cache
+stores the tuple of `(transaction_id, transaction_fingerprint_id)`.
+`transaction_id` is a UUID with 16-byte storage overhead and
+`transaction_fingerprint_id` is a unsigned 64-bit integer with 8-byte storage
+overhead. Hence, with an overhead of 1 MiB, we will be able to store ~43,690
+recently executed transactions. The total storage limit can be adjusted via
+cluster setting.
+
+Each node will expose a new RPC endpoint which allows for other nodes to query
+transaction fingerprint ID using transaction IDs.
+
+``` proto 
+message TransactionIDResolutionRequest {
+  repeated bytes transactionIDs = 1 [(gogoproto.customname) = "ID",
+      (gogoproto.customtype) = "github.com/cockroachdb/cockroach/pkg/util/uuid.UUID",
+      (gogoproto.nullable) = false];
+}
+
+message TransactionIDResolutionResponse {
```

Comment:
Would we perhaps use a nested message for these pairings; or even a [map](https://developers.google.com/protocol-buffers/docs/overview#maps)?

Reply:
Good call for using the nested message. `map` is unfortunately not an option for us because `TransactionID` (the key) has UUID type, which is represented as `bytes` in protobuf. Protobuf disallow using `bytes` in the key for `map`